### PR TITLE
[Search pipelines] Pass "adhocness" flag to processor factories

### PR DESCRIPTION
**Pull request by msfroh**
### Description
A named search pipeline may be created with a PUT request, while an "anonymous" or "ad hoc" search pipeline can be defined in the search request source. In the latter case, we don't want to create any "resource-heavy" processors, since they're potentially increasing the cost of every search request, whereas named pipeline processors get reused.

This change passes a configuration flag to a processor factory if it's being called as part of an ad hoc pipeline. The factory can use that information to avoid allocating expensive resources (maybe by throwing an exception instead).

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/8163

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
This feels a little overloaded and odd to me, hiding a lot of "internal mechanics" under a boolean. Is there a version in which we can express "a search pipeline defined in the search request source" in a more future-compatible way, e.g. as "source" of where the pipeline was defined?

File: CHANGELOG.md [hunk 1]
```diff
@@ -128,6 +128,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Compress and cache cluster state during validate join request ([#7321](https://github.com/opensearch-project/OpenSearch/pull/7321))
 - [Snapshot Interop] Add Changes in Create Snapshot Flow for remote store interoperability. ([#7118](https://github.com/opensearch-project/OpenSearch/pull/7118))
 - Allow insecure string settings to warn-log usage and advise to migration of a newer secure variant ([#5496](https://github.com/opensearch-project/OpenSearch/pull/5496))
+- [Search Pipelines] Pass "adhocness" flag to processor factories ([#8164](https://github.com/opensearch-project/OpenSearch/pull/8164))
```
Comment (dblock) [inline CHANGELOG.md:131]: I think you can just skip changelog for this one, but if you want to leave it, replace adhocness with what we are actually doing, i.e. pipeline source.

**Review by dblock — CHANGES_REQUESTED** [review]
Ok, this is a lot better. Continuing on this thought though, what's the most future-compatible interface? You say "the context for pipeline creation", so should this be `PinelineContext` and have a field called `source`?

**Comment by msfroh** [comment]
> Ok, this is a lot better. Continuing on this thought though, what's the most future-compatible interface? You say "the context for pipeline creation", so should this be PipelineContext and have a field called source?

That's a good idea. 

If we go that way, I'm going to go ahead and change the `create` method signature to pass it to the processor factory explicitly, rather than "hiding" it in the config map (which is inherently more "dynamic"). I had been trying to avoid changing the signature as a gesture that we had a stable API in time for 2.8, but this sounds like the kind of thing where we benefit from sitting behind the "experimental" label.

**Comment by macohen** [comment]
@dblock is this change acceptable? looking to merge for the 2.9 release.

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
Looks like it will need a manual backport, @msfroh

**Comment by mingshl** [comment]
Tried to help backport this, but it looks like it's blocking by this [PR](https://github.com/opensearch-project/OpenSearch/pull/7283). We need [#7283](https://github.com/opensearch-project/OpenSearch/pull/7283) getting to 2.x first then backport this one.

**Comment by msfroh** [comment]
> Tried to help backport this, but it looks like it's blocking by this https://github.com/opensearch-project/OpenSearch/pull/7283. We need https://github.com/opensearch-project/OpenSearch/pull/7283 getting to 2.x first then backport this one.

Looks like the backport (https://github.com/opensearch-project/OpenSearch/pull/8512) was just merged a few hours ago.

**Comment by msfroh** [comment]
Backport PR: https://github.com/opensearch-project/OpenSearch/pull/8522

**Pull request merged**